File: Wallet.js
class Wallet {
    constructor() {
        this.solde = 0.0;
        this.history = [];
    }

    addMoney(amount) {
        this.solde += amount;
        this.saveTransaction(`Ajout d'argent : +${amount} Ar`);
        console.log(`Argent ajouté au portefeuille. Solde actuel : ${this.solde} Ar`);
    }

    withdrawMoney(amount) {
        if (amount <= this.solde) {
            this.solde -= amount;
            this.saveTransaction(`Retrait d'argent : -${amount} Ar`);
            console.log(`Argent retiré du portefeuille. Solde actuel : ${this.solde} Ar`);
        } else {
            console.log("Solde insuffisant.");
        }
    }

    showBalance() {
        console.log(`Solde actuel du portefeuille : ${this.solde} Ar`);
    }

    calculateInterest(rateInterest) {
        const interest = this.solde * (rateInterest / 100);
        this.solde += interest;
        this.saveTransaction(`Calcul des intérêts : +${interest} Ar`);
        console.log(`Intérêts calculés. Nouveau solde : ${this.solde} Ar`);
    }

    showHistory() {
        console.log("\nHistorique des transactions : ");
        this.history.forEach(transaction => console.log(transaction));
    }

    saveTransaction(transaction) {
        this.history.push(transaction);
    }
}

const wallet = new Wallet();
const readlineSync = require('readline-sync');

console.log("Bienvenue dans votre portefeuille !");

let choice;

do {
    console.log("\n1. Ajouter de l'argent");
    console.log("2. Retirer de l'argent");
    console.log("3. Afficher le solde");
    console.log("4. Calculer des intérêts");
    console.log("5. Afficher l'historique des transactions");
    console.log("6. Quitter");
    choice = parseInt(readlineSync.question('Choix : '));

    switch (choice) {
        case 1:
            const amountAdd = parseFloat(readlineSync.question('Montant à ajouter : '));
            wallet.addMoney(amountAdd);
            break;
        case 2:
            const amountWithdrawal = parseFloat(readlineSync.question('Montant à retirer : '));
            wallet.withdrawMoney(amountWithdrawal);
            break;
        case 3:
            wallet.showBalance();
            break;
        case 4:
            const rateInterest = parseFloat(readlineSync.question('Taux d\'intérêt (%) : '));
            wallet.calculateInterest(rateInterest);
            break;
        case 5:
            wallet.showHistory();
            break;
        case 6:
            console.log("Au revoir ! Merci d'avoir utilisé votre portefeuille.");
            break;
        default:
            console.log("Choix non valide.");
    }
} while (choice !== 6);

File: Wallet.py
class Wallet:
    def __init__(self):
        self.solde = 0.0
        self.history = []

    def add_money(self, amount):
        self.solde += amount
        self.save_transaction(f"Ajout d'argent : +{amount} Ar")
        print(f"Argent ajouté au portefeuille. Solde actuel : {self.solde} Ar")

    def withdraw_money(self, amount):
        if amount <= self.solde:
            self.solde -= amount
            self.save_transaction(f"Retrait d'argent : -{amount} Ar")
            print(f"Argent retiré du portefeuille. Solde actuel : {self.solde} Ar")
        else:
            print("Solde insuffisant.")

    def show_balance(self):
        print(f"Solde actuel du portefeuille : {self.solde} Ar")

    def calculate_interest(self, rate_interest):
        interest = self.solde * (rate_interest / 100)
        self.solde += interest
        self.save_transaction(f"Calcul des intérêts : +{interest} Ar")
        print(f"Intérêts calculés. Nouveau solde : {self.solde} Ar")

    def show_history(self):
        print("\nHistorique des transactions : ")
        for transaction in self.history:
            print(transaction)

    def save_transaction(self, transaction):
        self.history.append(transaction)

wallet = Wallet()
print("Bienvenue dans votre portefeuille !")

while True:
    print("\n1. Ajouter de l'argent")
    print("2. Retirer de l'argent")
    print("3. Afficher le solde")
    print("4. Calculer des intérêts")
    print("5. Afficher l'historique des transactions")
    print("6. Quitter")
    choice = int(input("Choix : "))

    if choice == 1:
        amount_add = float(input("Montant à ajouter : "))
        wallet.add_money(amount_add)
    elif choice == 2:
        amount_withdrawal = float(input("Montant à retirer : "))
        wallet.withdraw_money(amount_withdrawal)
    elif choice == 3:
        wallet.show_balance()
    elif choice == 4:
        rate_interest = float(input("Taux d'intérêt (%) : "))
        wallet.calculate_interest(rate_interest)
    elif choice == 5:
        wallet.show_history()
    elif choice == 6:
        print("Au revoir ! Merci d'avoir utilisé votre portefeuille.")
        break
    else:
        print("Choix non valide.")
